mars/services/task/execution/ray/executor.py

```diff
@@ -118,29 +124,64 @@ def execute_subtask(
     subtask_id: str,
     subtask_chunk_graph: ChunkGraph,
     output_meta_keys: Set[str],
-    input_keys: List[str],
     *inputs,
 ):
-    logger.info("Begin to execute subtask: %s", subtask_id)
+    """The function used for execute subtask in ray task.
+    If subtask is shuffle mapper, no chunk meta will be returned, otherwise return chunk mata."""
     ensure_coverage()
     subtask_chunk_graph = deserialize(*subtask_chunk_graph)
-    # inputs = [i[1] for i in inputs]
+    logger.info("Begin to execute subtask: %s", subtask_id)
+    # optimize chunk graph.
+    subtask_chunk_graph = _optimize_subtask_graph(subtask_chunk_graph)
+    start_chunks = list(subtask_chunk_graph.iter_indep())
+    maybe_mapper_chunk = subtask_chunk_graph.result_chunks[0]
+    is_mapper = (
+        isinstance(maybe_mapper_chunk.op, MapReduceOperand)
+        and maybe_mapper_chunk.op.stage == OperandStage.map
+    )
+    if isinstance(start_chunks[0].op, FetchShuffle):
+        assert len(start_chunks) == 1, start_chunks
+        # the subtask is a reducer subtask
+        n_mappers = len(inputs)
+        # some reducer may have multiple output chunks, see `PSRSshuffle._execute_reduce` and
+        # https://user-images.githubusercontent.com/12445254/168569524-f09e42a7-653a-4102-bdf0-cc1631b3168d.png
+        reducer_chunks = subtask_chunk_graph.successors(start_chunks[0])
+        reducer_operands = set(c.op for c in reducer_chunks)
+        if len(reducer_operands) != 1:
+            raise ValueError(
+                f"Subtask {subtask_id} has more than 1 reduce operands: {subtask_chunk_graph.to_dot()}"
+            )
+        reducer_operand = reducer_chunks[0].op
+        reducer_index = reducer_operand.reducer_index
+        # mock input keys, keep this in sync with `MapReducerOperand#_iter_mapper_key_idx_pairs`
+        input_keys = [(i, reducer_index) for i in range(n_mappers)]
+    else:
+        input_keys = [c.key for c in start_chunks if isinstance(c.op, Fetch)]
     context = RayExecutionWorkerContext(
         lambda: RayTaskState.get_handle(task_id), zip(input_keys, inputs)
     )
-    # optimize chunk graph.
-    subtask_chunk_graph = _optimize_subtask_graph(subtask_chunk_graph)
-    # from data_key to results
+
     for chunk in subtask_chunk_graph.topological_iter():
         if chunk.key not in context:
-            execute(context, chunk.op)
+            wrapped_execute = log_exception_wrapper(
+                execute,
+                "Execute operand %s of graph %s failed.",
+                chunk.op,
+                subtask_chunk_graph.to_dot(),
```

Comment:
`subtask_chunk_graph.to_dot()` every execute is very slow. We can pass a lambda instead of string formatter to the `log_exception_wrapper`.

Reply:
Also, `log_exception_wrapper` every loop is very slow, too. Please wrap the entire loop or just wrap the `execute_subtask`.

Reply:
reverted to print exception directly